refactor(api): ot3: add gripper critical points

sfoster1 opened the pull request:
For now, add some gripper-specific critical points to the enum that the
gripper handles, and use them to calculate position offsets for motion.
JAW_CENTER is the default, and will be what gets used in most calls when
a gripper is present; the calibration points will be used for calibration.

These new critical points will cause an error if used with pipettes, and the pipette critical points will cause an error if used with the gripper.

It's no longer valid to move the mount critical point for the gripper, since that criticalpoint is attached directly to the carriage.

These values are tested on a proto build.

sfoster1 committed:
refactor(api): Add critical points to gripper

sfoster1 committed:
more gripper offset changes

sfoster1 committed:
same here

sfoster1 committed:
fix s-d makefile deps for python dist

sfoster1 committed:
refactor(api): add critical point handling for gripper

For now, add some gripper-specific critical points to the enum that the
gripper handles, and use them to calculate position offsets for motion.
JAW_CENTER is the default, and will be what gets used in most calls when
a gripper is present; the calibration points will be used for calibration.

sfoster1 committed:
now that's the right number

sfoster1 committed:
need more juice for the gripper z

Laura-Danielle commented on api/src/opentrons/hardware_control/instruments/gripper.py:
```diff
@@ -37,6 +37,16 @@ def __init__(
+        self._jaw_center_offset = (
+            Point(*self._config.jaw_center_offset_from_base) + base_offset
+        )
+        self._front_pin_offset = (
```
can we call this `_front_calibration_pin_offset` and `_back_calibration_pin_offset` respectively?

Laura-Danielle replied:
Also if we're only ever going to use one pin to calibrate, should we decide which one and only track that offset? Is the gripper going to be redesigned to only have 1 pin?

sfoster1 replied:
we should keep these until we bring up the actual calibration process.

sfoster1 replied:
https://github.com/Opentrons/opentrons/pull/11024/commits/8d885d7f236dd05ef2078e075a9e7ed0dd10c58e

Laura-Danielle commented on api/src/opentrons/hardware_control/instruments/gripper.py:
```diff
@@ -73,12 +83,15 @@ def update_calibration_offset(self, cal_offset: GripperCalibrationOffset) -> Non
         """
-        # TODO: add critical point implementation
-        return Point(0, 0, 0)
+        if cp_override == CriticalPoint.GRIPPER_FRONT_CALIBRATION_PIN:
```
If we do decide to just pick one location, can we call this: `CriticalPoint.GRIPPER_CALIBRATION`?

sfoster1 replied:
yeah

Laura-Danielle commented on api/src/opentrons/hardware_control/instruments/gripper.py:
```diff
@@ -37,6 +37,16 @@ def __init__(
+        self._back_pin_offset = (
+            Point(*self._config.pin_two_offset_from_base) + base_offset
+        )
         self._calibration_offset = gripper_cal_offset
```
Can we change this to: `calibrated_point` or something? Makes me think that this is a default configuration rather than a value we got from calibrating the thing.

sfoster1 replied:
it's not a point though, it's the offset you get from calibration that's added into the rest of the critical point offsets. i'm not sure what else to call it, but maybe comments would help?

sfoster1 replied:
https://github.com/Opentrons/opentrons/pull/11024/commits/8d885d7f236dd05ef2078e075a9e7ed0dd10c58e

Laura-Danielle replied:
hmm comments are fine for now, I can't think of a better name at the moment other than maybe `_mount_to_calibration_pin_offset` which is kinda long lol.

Laura-Danielle commented on api/src/opentrons/hardware_control/instruments/gripper_handler.py:
```diff
@@ -47,10 +47,10 @@ def gripper(self, gripper: Optional[Gripper] = None) -> None:
-        if not self._gripper:
-            return Point(0, 0, 0)
-        else:
+        if self._gripper and cp_override != CriticalPoint.MOUNT:
```
Discussed in a DM, but we shouldn't have a `Mount` critical point for the gripper since the mount w/o a gripper attached cannot move independently of the XY axes and also doesn't have Z movement.

sfoster1 replied:
https://github.com/Opentrons/opentrons/pull/11024/commits/29688aa156fc4dfd3c73a692954b6b19d923c9b6

Laura-Danielle reviewed (commented):
Initial thoughts about naming for offsets. We can push these decisions to later as well, or to a wider group but if we're only ever using one pin to calibrate then we should think about reducing the number of sensors as well as only have 1 pin available to add to the gripper in the mechanical design -- otherwise I could see it getting confusing for users.

sfoster1 committed:
Critical point validity checking

- It's no longer valid to use the MOUNT cp on a gripper
- There are actual exceptions for using a gripper cp on a pipette and
- vice versa

sfoster1 committed:
whoops

sfoster1 committed:
change pin names, add comments

sfoster1 committed:
fixups from rebase

Laura-Danielle reviewed (approved):
lgtm
